pyteomics/proforma.py

```diff
@@ -0,0 +1,1580 @@
+        defn = None
+        for resolver in self.resolvers:
+            try:
+                defn = resolver(name=name, id=id, **kwargs)
+            except (KeyError):
+                continue
+        if defn is None:
+            if name is None:
+                raise KeyError(id)
+            elif id is None:
+                raise KeyError(name)
+            else:
+                raise ValueError("Must provide one of `name` or `id`")
+        return defn
+
+
+class ModificationBase(TagBase):
+    '''A base class for all modification tags with marked prefixes.
+    '''
+
+    _tag_type = None
+    __slots__ = ('_definition', )
+
+    def __init__(self, value, extra=None, group_id=None):
+        super(ModificationBase, self).__init__(
+            self._tag_type, value, extra, group_id)
+        self._definition = None
+
+    @property
+    def definition(self):
+        if self._definition is None:
+            self._definition = self.resolve()
+        return self._definition
+
+    @property
+    def mass(self):
+        return self.definition['mass']
+
+    @property
+    def composition(self):
+        return self.definition.get('composition')
+
+    @property
+    def id(self):
+        return self.definition.get('id')
+
+    @property
+    def name(self):
+        return self.definition.get('name')
+
+    @property
+    def provider(self):
+        return self.definition.get('provider')
+
+    def _populate_from_definition(self, definition):
+        self._definition = definition
+
+    def _format_main(self):
+        return "{self.prefix_name}:{self.value}".format(self=self)
+
+    def _parse_identifier(self):
+        tokens = self.value.split(":", 1)
+        if len(tokens) > 1:
+            value = tokens[1]
+        else:
+            value = self.value
+        if value.isdigit():
+            id = int(value)
+            name = None
+        else:
+            name = value
+            id = None
+        return name, id
+
+    def resolve(self):
+        '''Find the term and return it's properties
+        '''
+        keys = self._parse_identifier()
+        return self.resolver(*keys)
+
+
+class FormulaModification(ModificationBase):
+    prefix_name = "Formula"
+
+    isotope_pattern = re.compile(r'\[(?P<isotope>\d+)(?P<element>[A-Z][a-z]*)(?P<quantity>[\-+]?\d+)\]')
+    _tag_type = TagTypeEnum.formula
+
+    def _normalize_isotope_notation(self, match):
+        '''Rewrite ProForma isotope notation to Pyteomics-compatible
+        isotope notation.
+
+        Parameters
+        ----------
+        match : Match
+            The matched isotope notation string parsed by the regular expression.
+
+        Returns
+        reformatted : str
+            The re-written isotope notation
+        '''
+        parts = match.groupdict()
+        return "{element}[{isotope}]{quantity}".format(**parts)
+
+    def resolve(self):
+        normalized = ''.join(self.value.split(" "))
```

Comment:
Is this different from `self.value.replace(' ', '')`?

Reply:
Your way is more efficient. This is a left-over from my thinking about operating on each space-separated piece separately, but ultimately it wasn't necessary. I think the provision for spaces in formulae are an artefact of history now.